<a href="https://colab.research.google.com/github/KatBotkowska/Pandas-Numpy-exercices/blob/master/Selecting_subsets_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## How to select subsets in Pandas - part 2: boolean indexing
*Przy wybieraniu za pomocą: [], .loc, .iloc - wybiera się na podstawie nagłówków.
Przy wybieraniu przy pomocy boolean indexing - na podstawie wartości danych., które przyrównuje sie do pytania - True lub False*

In [0]:
import pandas as pd
import numpy as np

In [0]:
path = 'https://raw.githubusercontent.com/tdpetrou/Learn-Pandas/master/data/stackoverflow_qa.csv'
stof = pd.read_csv(path)

In [5]:
stof_head = stof.head()
stof_head

id         creationdate  score  ...  quest_rep         ans_name  ans_rep
0  5486226  2011-03-30 12:26:50      4  ...      125.0  Mike Pennington  26995.0
1  5515021  2011-04-01 14:50:44      8  ...     3301.0  Mike Pennington  26995.0
2  5558607  2011-04-05 21:13:50      2  ...     3301.0     Wes McKinney  43310.0
3  6467832  2011-06-24 12:31:45      9  ...      117.0     Wes McKinney  43310.0
4  7577546  2011-09-28 01:58:38      9  ...      958.0             HYRY  54137.0

[5 rows x 12 columns]

In [6]:
criteria = [True, False, False, True, False] #wybierze wiersze dla których wskazano True: 0, 3
stof_head[criteria]

id         creationdate  score  ...  quest_rep         ans_name  ans_rep
0  5486226  2011-03-30 12:26:50      4  ...      125.0  Mike Pennington  26995.0
3  6467832  2011-06-24 12:31:45      9  ...      117.0     Wes McKinney  43310.0

[2 rows x 12 columns]

In [8]:
s = pd.Series([True, False, False, True, False]) #w Pandas używa się Series jako podstawowego kontenera
stof_head[s]

id         creationdate  score  ...  quest_rep         ans_name  ans_rep
0  5486226  2011-03-30 12:26:50      4  ...      125.0  Mike Pennington  26995.0
3  6467832  2011-06-24 12:31:45      9  ...      117.0     Wes McKinney  43310.0

[2 rows x 12 columns]

In [9]:
t = np.array([True, False, False, True, False]) #wrzucenie listy boolean do array
stof_head[t]

id         creationdate  score  ...  quest_rep         ans_name  ans_rep
0  5486226  2011-03-30 12:26:50      4  ...      125.0  Mike Pennington  26995.0
3  6467832  2011-06-24 12:31:45      9  ...      117.0     Wes McKinney  43310.0

[2 rows x 12 columns]

*w przypadku indeksowania za pomocą listy, serii wartości True lub False, trzeba uważać by lista wartości boolean była równa ilosci wierszy tabeli. Jeśli nie - IndexingError.
ALE: nie używa się list wpisywanych 'z ręki' do indeksowania, tylko operatorów porównania: <, >, =...*

## Indeksowanie boolean - pojedynczy warunek

### Wskazanie kryteriów jako listy boolean dla pewnego warunku

In [10]:
criteria = stof['score'] >= 10 #tworzy kryterium, dla którego score jest większy niż 10
criteria.head() #zwraca listę wartości boolean dla pozycji kolumny score, które spełniają warunek

0    False
1    False
2    False
3    False
4    False
Name: score, dtype: bool

In [11]:
stof_score_10_or_more = stof[criteria] #wybiera wszystkie wiersze, dla których spełniony w kolumnie score jest warunek
stof_score_10_or_more.head()

id         creationdate  score  ...  quest_rep       ans_name  ans_rep
6   7776679  2011-10-15 08:21:17     25  ...      882.0   Wes McKinney  43310.0
7   7813132  2011-10-18 20:16:12     10  ...      882.0   Wes McKinney  43310.0
8   7837722  2011-10-20 14:46:14    201  ...     1563.0  Nick Crawford   2779.0
14  8916302  2012-01-18 19:41:27     29  ...    19244.0   Wes McKinney  43310.0
17  8991709  2012-01-24 17:59:53    136  ...    12484.0     Matt Dowle  41275.0

[5 rows x 12 columns]

In [12]:
stof.shape

(56398, 12)

In [13]:
stof_score_10_or_more.shape

(1505, 12)

*Z ponad 56tys. wierszy, warunek dla score spełnia 1505 wierszy*

### Boolean selection in one line

In [15]:
stof[stof['score'] >= 10].head() #zapisanie powyższych warunków w jednej linii

id         creationdate  score  ...  quest_rep       ans_name  ans_rep
6   7776679  2011-10-15 08:21:17     25  ...      882.0   Wes McKinney  43310.0
7   7813132  2011-10-18 20:16:12     10  ...      882.0   Wes McKinney  43310.0
8   7837722  2011-10-20 14:46:14    201  ...     1563.0  Nick Crawford   2779.0
14  8916302  2012-01-18 19:41:27     29  ...    19244.0   Wes McKinney  43310.0
17  8991709  2012-01-24 17:59:53    136  ...    12484.0     Matt Dowle  41275.0

[5 rows x 12 columns]

In [16]:
criteria = stof['ans_name'] == 'Scott Boston' #kryterium, by nazwa odpowiadającego to Scott Boston
criteria.head()

0    False
1    False
2    False
3    False
4    False
Name: ans_name, dtype: bool

In [17]:
stof[criteria].head()

id         creationdate  score  ...  quest_rep      ans_name  ans_rep
38161  43491342  2017-04-19 09:14:28      4  ...     3991.0  Scott Boston  23611.0
38178  43190850  2017-04-03 17:31:33      1  ...      545.0  Scott Boston  23611.0
38237  43176052  2017-04-03 03:21:12      2  ...    31923.0  Scott Boston  23611.0
38246  43209525  2017-04-04 14:03:17      5  ...     5470.0  Scott Boston  23611.0
38275  43211893  2017-04-04 15:45:17      0  ...     1102.0  Scott Boston  23611.0

[5 rows x 12 columns]

In [18]:
stof[stof['ans_name']== 'Scott Boston'].head()

id         creationdate  score  ...  quest_rep      ans_name  ans_rep
38161  43491342  2017-04-19 09:14:28      4  ...     3991.0  Scott Boston  23611.0
38178  43190850  2017-04-03 17:31:33      1  ...      545.0  Scott Boston  23611.0
38237  43176052  2017-04-03 03:21:12      2  ...    31923.0  Scott Boston  23611.0
38246  43209525  2017-04-04 14:03:17      5  ...     5470.0  Scott Boston  23611.0
38275  43211893  2017-04-04 15:45:17      0  ...     1102.0  Scott Boston  23611.0

[5 rows x 12 columns]

## Multiple conditions

**&  and**
**|  or**
**~   not**

**AND  &**

*Wiersze, dla których score >=5, a autorem odpowiedzi był Scott Boston*

### 1 sposób - przypisanie warunku do listy boolean

In [19]:
condition1 = stof['score']>=5
condition2 = stof['ans_name'] == 'Scott Boston'
condition_all = condition1 & condition2
stof[condition_all].head()

id         creationdate  score  ...  quest_rep      ans_name  ans_rep
38246  43209525  2017-04-04 14:03:17      5  ...     5470.0  Scott Boston  23611.0
38640  42870703  2017-03-18 05:06:22      5  ...       43.0  Scott Boston  23611.0
44358  45064916  2017-07-12 18:16:49      5  ...      506.0  Scott Boston  23611.0
44814  44877663  2017-07-03 04:07:59      9  ...      181.0  Scott Boston  23611.0
52013  47061564  2017-11-01 18:40:36      5  ...     1890.0  Scott Boston  23611.0

[5 rows x 12 columns]

### 2gi sposób - w jednej linii
warunki wrzucone do () i połączone &

In [21]:
stof[(stof['score']>=5) & (stof['ans_name'] == 'Scott Boston')].head()

id         creationdate  score  ...  quest_rep      ans_name  ans_rep
38246  43209525  2017-04-04 14:03:17      5  ...     5470.0  Scott Boston  23611.0
38640  42870703  2017-03-18 05:06:22      5  ...       43.0  Scott Boston  23611.0
44358  45064916  2017-07-12 18:16:49      5  ...      506.0  Scott Boston  23611.0
44814  44877663  2017-07-03 04:07:59      9  ...      181.0  Scott Boston  23611.0
52013  47061564  2017-11-01 18:40:36      5  ...     1890.0  Scott Boston  23611.0

[5 rows x 12 columns]

**OR  |**

*pytania, dla których punktacja była co najmniej 100, a odpowiedzi było co najmniej 10*

In [23]:
condition1 = stof['score']>=100
condition2 = stof['answercount']>=10
condition_all = condition1 | condition2
stof[condition_all].head()

id         creationdate  score  ...  quest_rep       ans_name  ans_rep
8    7837722  2011-10-20 14:46:14    201  ...     1563.0  Nick Crawford   2779.0
17   8991709  2012-01-24 17:59:53    136  ...    12484.0     Matt Dowle  41275.0
60  10065051  2012-04-08 18:01:13     83  ...      532.0            NaN      NaN
75  10373660  2012-04-29 16:10:35    199  ...     2421.0   Wes McKinney  43310.0
99  10636024  2012-05-17 12:48:00     36  ...     1562.0    user1319128    555.0

[5 rows x 12 columns]

In [24]:
stof[(stof['score']>=100) | (stof['answercount']>=10)].head()

id         creationdate  score  ...  quest_rep       ans_name  ans_rep
8    7837722  2011-10-20 14:46:14    201  ...     1563.0  Nick Crawford   2779.0
17   8991709  2012-01-24 17:59:53    136  ...    12484.0     Matt Dowle  41275.0
60  10065051  2012-04-08 18:01:13     83  ...      532.0            NaN      NaN
75  10373660  2012-04-29 16:10:35    199  ...     2421.0   Wes McKinney  43310.0
99  10636024  2012-05-17 12:48:00     36  ...     1562.0    user1319128    555.0

[5 rows x 12 columns]

**NOT ~**
*pytania z punktacją większą niż 100*

In [27]:
stof[~(stof['score']<=100)].head()

id         creationdate  score  ...  quest_rep       ans_name  ans_rep
8     7837722  2011-10-20 14:46:14    201  ...     1563.0  Nick Crawford   2779.0
17    8991709  2012-01-24 17:59:53    136  ...    12484.0     Matt Dowle  41275.0
75   10373660  2012-04-29 16:10:35    199  ...     2421.0   Wes McKinney  43310.0
100  10665889  2012-05-19 14:11:42    144  ...      988.0     Ted Petrou  10426.0
106  10715965  2012-05-23 08:12:31    340  ...     1988.0           fred   2342.0

[5 rows x 12 columns]

## Complex condition

*pytania odpowiedziane przez Scott Boston z punktacją 5 i więcej lub odpowiedziane przez Ted Petrou z punktacją 5 i więcej*
**Warunki muszą być w nawiasach**

In [28]:
stof[((stof['ans_name'] == 'Scott Boston') & (stof['score']>=5)) | ((stof['ans_name'] == 'Ted Petrou') & (stof['score']>=5))].head()

id         creationdate  score  ...  quest_rep      ans_name  ans_rep
100    10665889  2012-05-19 14:11:42    144  ...      988.0    Ted Petrou  10426.0
38246  43209525  2017-04-04 14:03:17      5  ...     5470.0  Scott Boston  23611.0
38640  42870703  2017-03-18 05:06:22      5  ...       43.0  Scott Boston  23611.0
44358  45064916  2017-07-12 18:16:49      5  ...      506.0  Scott Boston  23611.0
44386  44635626  2017-06-19 16:28:34      7  ...       92.0    Ted Petrou  10426.0

[5 rows x 12 columns]

**isin  - many condition with or in one column**

*pytania odpowiedziane przez Scott Boston, Ted Petrou, MaxU, unutbu*

wersja 1

In [29]:
criteria = ((stof['ans_name'] == 'Scott Boston') | (stof['ans_name'] == 'Ted Petrou') | (stof['ans_name'] == 'MaxU') | (stof['ans_name'] == 'unutbu'))
stof[criteria].head()

id         creationdate  score  ...  quest_rep    ans_name   ans_rep
100  10665889  2012-05-19 14:11:42    144  ...      988.0  Ted Petrou   10426.0
574  13385860  2012-11-14 19:25:28     29  ...      530.0      unutbu  464745.0
593  13446480  2012-11-19 01:20:07     18  ...     8714.0      unutbu  464745.0
599  13460889  2012-11-19 19:39:06      8  ...     5013.0      unutbu  464745.0
639  13653030  2012-11-30 20:54:35      6  ...      323.0      unutbu  464745.0

[5 rows x 12 columns]

wersja 2

In [30]:
criteria = stof['ans_name'].isin(['Scott Boston', 'Ted Petrou', 'MaxU', 'unutbu'])
stof[criteria].head()

id         creationdate  score  ...  quest_rep    ans_name   ans_rep
100  10665889  2012-05-19 14:11:42    144  ...      988.0  Ted Petrou   10426.0
574  13385860  2012-11-14 19:25:28     29  ...      530.0      unutbu  464745.0
593  13446480  2012-11-19 01:20:07     18  ...     8714.0      unutbu  464745.0
599  13460889  2012-11-19 19:39:06      8  ...     5013.0      unutbu  464745.0
639  13653030  2012-11-30 20:54:35      6  ...      323.0      unutbu  464745.0

[5 rows x 12 columns]

In [32]:
stof[stof['ans_name'].isin(['Scott Boston', 'Ted Petrou', 'MaxU', 'unutbu'])].head()

id         creationdate  score  ...  quest_rep    ans_name   ans_rep
100  10665889  2012-05-19 14:11:42    144  ...      988.0  Ted Petrou   10426.0
574  13385860  2012-11-14 19:25:28     29  ...      530.0      unutbu  464745.0
593  13446480  2012-11-19 01:20:07     18  ...     8714.0      unutbu  464745.0
599  13460889  2012-11-19 19:39:06      8  ...     5013.0      unutbu  464745.0
639  13653030  2012-11-30 20:54:35      6  ...      323.0      unutbu  464745.0

[5 rows x 12 columns]

*Pytania odpowiedziane przez wyżej plus punkty wiecej niż 30*

In [33]:
stof[(stof['ans_name'].isin(['Scott Boston', 'Ted Petrou', 'MaxU', 'unutbu']))& (stof['score']>30)].head()

id         creationdate  score  ...  quest_rep    ans_name   ans_rep
100   10665889  2012-05-19 14:11:42    144  ...      988.0  Ted Petrou   10426.0
680   13842088  2012-12-12 14:40:45    160  ...      902.0      unutbu  464745.0
977   15465645  2013-03-17 20:12:23     36  ...     9617.0      unutbu  464745.0
1253  17071871  2013-06-12 17:42:05    475  ...     3374.0      unutbu  464745.0
1296  16522380  2013-05-13 12:44:17     56  ...     2406.0      unutbu  464745.0

[5 rows x 12 columns]

### isnull - sprawdzanie gdzie są missing values

*pytania, gdzie nie ma zaakceptowanej odpowiedzi, czyli ans_name - jest puste, bo nie ma autora zaakceptowanej odpowiedzi*

In [34]:
no_answer = stof['ans_name'].isnull()
stof[no_answer].head()

id         creationdate  score  ...  quest_rep ans_name  ans_rep
5   7766400  2011-10-14 10:33:54      2  ...      100.0      NaN      NaN
11  8451327  2011-12-09 20:27:24      1  ...        6.0      NaN      NaN
16  8966871  2012-01-23 03:21:00      0  ...       16.0      NaN      NaN
29  9641916  2012-03-09 22:36:52      7  ...       41.0      NaN      NaN
30  9647656  2012-03-10 15:35:26      3  ...       45.0      NaN      NaN

[5 rows x 12 columns]

In [36]:
stof[stof['ans_name'].isnull()].tail()

id         creationdate  score  ...  quest_rep ans_name  ans_rep
56391  47601533  2017-12-01 21:09:16      1  ...        6.0      NaN      NaN
56392  47610316  2017-12-02 17:02:03      1  ...        6.0      NaN      NaN
56394  47607549  2017-12-02 11:41:59     -1  ...        1.0      NaN      NaN
56395  47609325  2017-12-02 15:16:26      0  ...        1.0      NaN      NaN
56397  47611969  2017-12-02 19:59:09     -6  ...        1.0      NaN      NaN

[5 rows x 12 columns]

**Boolean selection on series**

In [37]:
s = stof['commentcount']
s.head()

0    4
1    6
2    0
3    0
4    0
Name: commentcount, dtype: int64

In [39]:
criteria = s > 10
s[criteria].head()

17     16
76     14
566    11
763    12
781    19
Name: commentcount, dtype: int64

In [40]:
s[s>10].head()

17     16
76     14
566    11
763    12
781    19
Name: commentcount, dtype: int64

In [41]:
s[(s>10)&(s<15)].head()

76     14
566    11
763    12
787    12
837    13
Name: commentcount, dtype: int64

In [42]:
s[s.between(11, 14)].head()

76     14
566    11
763    12
787    12
837    13
Name: commentcount, dtype: int64

## .loc - wybiera przez nagłówki, ale może też być do boolean indexing przez operatory porównania

In [44]:
stof.loc[(stof['ans_name'] == 'Scott Boston') & (stof['score']>=5)].head()

id         creationdate  score  ...  quest_rep      ans_name  ans_rep
38246  43209525  2017-04-04 14:03:17      5  ...     5470.0  Scott Boston  23611.0
38640  42870703  2017-03-18 05:06:22      5  ...       43.0  Scott Boston  23611.0
44358  45064916  2017-07-12 18:16:49      5  ...      506.0  Scott Boston  23611.0
44814  44877663  2017-07-03 04:07:59      9  ...      181.0  Scott Boston  23611.0
52013  47061564  2017-11-01 18:40:36      5  ...     1890.0  Scott Boston  23611.0

[5 rows x 12 columns]

In [45]:
stof.loc[stof['ans_name'].isin(['Scott Boston', 'Ted Petrou', 'MaxU', 'unutbu'])].head()

id         creationdate  score  ...  quest_rep    ans_name   ans_rep
100  10665889  2012-05-19 14:11:42    144  ...      988.0  Ted Petrou   10426.0
574  13385860  2012-11-14 19:25:28     29  ...      530.0      unutbu  464745.0
593  13446480  2012-11-19 01:20:07     18  ...     8714.0      unutbu  464745.0
599  13460889  2012-11-19 19:39:06      8  ...     5013.0      unutbu  464745.0
639  13653030  2012-11-30 20:54:35      6  ...      323.0      unutbu  464745.0

[5 rows x 12 columns]

## .loc select rows and columns
*Można kombinować boolean selection dla wierszy i wybierać kolumny nagłówkami*

*Pytania z oglądającymi więcej niż 20k ale tylko kolumny creationdate, viewcount, ans_name*

In [47]:
stof.loc[stof['viewcount']>20000, ['creationdate', 'viewcount', 'ans_name']].head()

creationdate  viewcount       ans_name
6   2011-10-15 08:21:17      28159   Wes McKinney
8   2011-10-20 14:46:14     223746  Nick Crawford
14  2012-01-18 19:41:27      20614   Wes McKinney
31  2012-03-11 06:00:56      39323           huon
35  2012-03-18 12:53:06      43262   Wes McKinney

*wiersze gdzie favoritecount jest pomiędzy 30 a 40, i kolumny co 3cią, zaczynając od title*

In [48]:
stof.loc[stof['favoritecount'].between(30, 40), 'title'::3].head()

title  ...     ans_name
348  Pandas: create two new columns in a dataframe ...  ...          DSM
555  How can I replace all the NaN values with Zero...  ...         Aman
643  Pandas DataFrame: remove unwanted parts from s...  ...       eumiro
667  Creating an empty Pandas DataFrame, then filli...  ...  Andy Hayden
892  Apply multiple functions to multiple groupby c...  ...     Zelazny7

[5 rows x 3 columns]

## Boolean selection for columns

In [49]:
stof.shape

(56398, 12)

In [0]:
col_bools =[True, False, False]*4 #przykładowa seria dla bool

In [51]:
stof.loc[:, col_bools].head()


id  viewcount  commentcount  quest_rep
0  5486226       2113             4      125.0
1  5515021       7015             6     3301.0
2  5558607       7392             0     3301.0
3  6467832      13056             0      117.0
4  7577546       2488             0      958.0

*wybieramy te same kolumny, ale wiersze gdzie liczba odsłon >500k*

In [52]:
stof.loc[stof['viewcount']>500000, col_bools].head()

id  viewcount  commentcount  quest_rep
171   11285613     526432             4     3369.0
181   11346283     931604             1     4206.0
397   12555323     698537             0     2920.0
581   13411544     802655             0     8807.0
1253  17071871     549481             0     3374.0

**EXAMPLE**

rzucamy 10 monetami 100 razy - wyniki do tabeli do kolumn dla każdej monety

In [53]:
coins = pd.DataFrame(np.random.randint(0,2, (100,10)), columns = list('abcdefghij'))
coins.head()

a  b  c  d  e  f  g  h  i  j
0  0  0  1  1  1  0  0  1  1  0
1  1  0  0  0  1  0  0  0  1  0
2  0  0  0  0  1  0  1  1  1  0
3  0  1  1  0  1  0  0  1  0  1
4  0  1  0  0  0  1  1  0  0  1

In [55]:
coins.shape

(100, 10)

*wybieramy kolumny, gdzie więcej niż 50% wyrzucono heads*

In [57]:
coin_mean = coins.mean()
coin_mean

a    0.45
b    0.47
c    0.53
d    0.57
e    0.50
f    0.49
g    0.55
h    0.42
i    0.46
j    0.52
dtype: float64

In [61]:
coins.loc[:, coin_mean>0.50].head()

c  d  g  j
0  1  1  0  0
1  0  0  0  0
2  0  0  1  0
3  1  0  0  1
4  0  0  1  1

**Porównywanie kolumn**

*kolumny z pytaniami, gdzie jest więcej odpowiedzi niż punktów*

In [62]:
criteria = stof['answercount'] > stof['score']
stof[criteria].head()

id         creationdate  ...                ans_name   ans_rep
10   8273092  2011-11-25 18:39:02  ...              codingknob    2279.0
46   9927711  2012-03-29 14:42:42  ...                  eumiro  104313.0
54  10003171  2012-04-03 23:59:41  ...            Wes McKinney   43310.0
59  10027719  2012-04-05 11:28:00  ...            Wes McKinney   43310.0
77  10393447  2012-05-01 04:12:13  ...  Ignacio Vazquez-Abrams  513959.0

[5 rows x 12 columns]

In [63]:
stof[stof['answercount']>stof['score']].head()

id         creationdate  ...                ans_name   ans_rep
10   8273092  2011-11-25 18:39:02  ...              codingknob    2279.0
46   9927711  2012-03-29 14:42:42  ...                  eumiro  104313.0
54  10003171  2012-04-03 23:59:41  ...            Wes McKinney   43310.0
59  10027719  2012-04-05 11:28:00  ...            Wes McKinney   43310.0
77  10393447  2012-05-01 04:12:13  ...  Ignacio Vazquez-Abrams  513959.0

[5 rows x 12 columns]

**Uwaga: przy bool nie używamy .iloc, bo używa się nazw kolumn do porównywania**

## Exercises

*pytania z ilością odpowiedzi = 5*

In [66]:
stof.loc[stof['answercount']==5].head()

id         creationdate  score  ...  quest_rep      ans_name   ans_rep
75   10373660  2012-04-29 16:10:35    199  ...     2421.0  Wes McKinney   43310.0
115  10791661  2012-05-29 00:06:52      7  ...      958.0        lbolla    4552.0
130  10972410  2012-06-10 21:12:43     19  ...      895.0      BrenBarn  136870.0
189  11391969  2012-07-09 09:04:33     44  ...     9605.0  Wes McKinney   43310.0
246  11811392  2012-08-04 19:25:34     17  ...     1260.0      BrenBarn  136870.0

[5 rows x 12 columns]

*pytania, gdzie ilość views mniej niż 10*

In [67]:
stof.loc[stof['viewcount']<10].head()

id         creationdate  score  ...  quest_rep ans_name  ans_rep
7787   26493306  2014-10-21 18:06:10      1  ...     3551.0      NaN      NaN
17653  33641540  2015-11-10 23:19:36      0  ...       88.0      NaN      NaN
29414  40062672  2016-10-15 18:24:21     -1  ...      186.0      NaN      NaN
36086  42538091  2017-03-01 17:20:37      0  ...      119.0      NaN      NaN
36340  42080039  2017-02-07 01:13:30      0  ...     2281.0     root  12202.0

[5 rows x 12 columns]

*pytania, gdzie osoba pytająca jest osobą odpowiadającą*

In [68]:
stof.loc[stof['quest_name'] == stof['ans_name']].head()

id         creationdate  score  ...  quest_rep     ans_name  ans_rep
10   8273092  2011-11-25 18:39:02      1  ...     2279.0   codingknob   2279.0
33   9721429  2012-03-15 14:08:31      9  ...     3369.0  user1234440   3369.0
58  10020591  2012-04-04 23:17:23     21  ...    17129.0          bmu  17129.0
68  10175068  2012-04-16 13:27:44     15  ...    19551.0       elyase  19551.0
74  10264739  2012-04-22 02:35:06      9  ...     1752.0     vgoklani   1752.0

[5 rows x 12 columns]

*pytanie, gdzie nie ma zaakceptowanej odpowiedzi, ale score jest większy niż 100*

In [69]:
stof.loc[(stof['ans_name'].isnull()) & (stof['score']>100)].head()

id         creationdate  score  ...  quest_rep ans_name  ans_rep
184   11350770  2012-07-05 18:57:34    149  ...      816.0      NaN      NaN
514   13148429  2012-10-30 22:22:59    312  ...     1689.0      NaN      NaN
527   13187778  2012-11-02 00:57:33    102  ...      511.0      NaN      NaN
560   13331698  2012-11-11 13:48:53    154  ...     7865.0      NaN      NaN
5013  19377969  2013-10-15 09:42:52    128  ...      831.0      NaN      NaN

[5 rows x 12 columns]

*pytania, gdzie reputacja pytającego jest większa niż osoby odpowiadającej. Potem procent jak często to sie zdarza*

In [70]:
stof.loc[stof['quest_rep']>stof['ans_rep']].head()

id         creationdate  ...           ans_name  ans_rep
26    9588331  2012-03-06 17:01:47  ...  Jeff Hammerbacher   3172.0
51    9962114  2012-04-01 05:25:20  ...        Josh Hemann    590.0
99   10636024  2012-05-17 12:48:00  ...        user1319128    555.0
101  10671227  2012-05-20 06:14:57  ...             Karmel   2870.0
112  10760364  2012-05-25 19:33:39  ...              minrk  22244.0

[5 rows x 12 columns]

In [71]:
stof.shape

(56398, 12)

In [73]:
stof.loc[stof['quest_rep']>stof['ans_rep']].shape

2240

In [75]:
rep_perc = (stof.loc[stof['quest_rep']>stof['ans_rep']].shape[0])/stof.shape[0]*100
rep_perc

3.9717720486542074

In [76]:
(stof['quest_rep']>stof['ans_rep']).mean() #prościej

0.039717720486542075

*pytanie, gdzie liczba odpowiedzi jest między 5 a 10, a liczba odsłon mniejsza niż 1k*

In [79]:
stof[stof['answercount'].between(5, 10) & (stof['viewcount']<1000)].head()

id         creationdate  score  ...  quest_rep  ans_name  ans_rep
620   13580875  2012-11-27 09:19:07      2  ...    15061.0  Victor K    332.0
1316  14742893  2013-02-07 03:13:31      2  ...      886.0       NaN      NaN
4816  21315997  2014-01-23 18:02:57      2  ...     1465.0      Jeff  62248.0
4958  19892107  2013-11-10 16:28:50      5  ...       32.0       NaN      NaN
5339  21040578  2014-01-10 09:33:09     -1  ...       16.0       NaN      NaN

[5 rows x 12 columns]

*Przeciwieństwo tego co wyżej*

In [80]:
stof[~(stof['answercount'].between(5, 10) & (stof['viewcount']<1000))].head()

id         creationdate  score  ...  quest_rep         ans_name  ans_rep
0  5486226  2011-03-30 12:26:50      4  ...      125.0  Mike Pennington  26995.0
1  5515021  2011-04-01 14:50:44      8  ...     3301.0  Mike Pennington  26995.0
2  5558607  2011-04-05 21:13:50      2  ...     3301.0     Wes McKinney  43310.0
3  6467832  2011-06-24 12:31:45      9  ...      117.0     Wes McKinney  43310.0
4  7577546  2011-09-28 01:58:38      9  ...      958.0             HYRY  54137.0

[5 rows x 12 columns]

**Inna baza danych**

In [81]:
path = 'https://raw.githubusercontent.com/tdpetrou/Learn-Pandas/6bb60eece5d6a1d2580351574ed5b6d0ee1a1dda/data/employee.csv'
employee = pd.read_csv(path)
employee.head()

POSITION_TITLE  ...    JOB_DATE
0  ASSISTANT DIRECTOR (EX LVL)  ...  2012-10-13
1            LIBRARY ASSISTANT  ...  2010-09-18
2               POLICE OFFICER  ...  2015-02-03
3            ENGINEER/OPERATOR  ...  1991-05-25
4                  ELECTRICIAN  ...  1994-10-22

[5 rows x 8 columns]

In [82]:
employee.RACE.value_counts()

Black or African American            700
White                                665
Hispanic/Latino                      480
Asian/Pacific Islander               107
American Indian or Alaskan Native     11
Others                                 2
Name: RACE, dtype: int64

*Find all the **`Black or African American`** females that work in the **`Houston Police Department-HPD`*

In [83]:
employee.loc[(employee['DEPARTMENT']=='Houston Police Department-HPD')&(employee['RACE']=='Black or African American') & (employee['GENDER']=='Female')].head()

POSITION_TITLE  ...    JOB_DATE
55   ADMINISTRATIVE ASSOCIATE  ...  2008-05-17
113     SENIOR POLICE OFFICER  ...  2007-02-24
137            JAIL ATTENDANT  ...  2013-07-17
196     SENIOR POLICE OFFICER  ...  2008-03-08
228     FINANCIAL ANALYST III  ...  2009-07-25

[5 rows x 8 columns]

*Find the females that have a salary over 100,000 OR males with salary under 50,000*

In [84]:
employee.loc[((employee['GENDER']=='Female')&(employee['BASE_SALARY']>100000)) | ((employee['GENDER']=='Male')&(employee['BASE_SALARY']<50000))].head()

POSITION_TITLE  ...    JOB_DATE
0         ASSISTANT DIRECTOR (EX LVL)  ...  2012-10-13
2                      POLICE OFFICER  ...  2015-02-03
7                           CARPENTER  ...  2013-11-04
9      AIRPORT OPERATIONS COORDINATOR  ...  2016-03-14
12  CUSTOMER SERVICE REPRESENTATIVE I  ...  2015-11-16

[5 rows x 8 columns]

*Find the females in the following departments with salary over 60,000 (Parks & Recreation, Solid Waste Management, Fleet Management Department, Library)*

In [87]:
employee.loc[employee['DEPARTMENT'].isin(['Parks&Recreation', 'Solid Waste Management', 'Fleet Management Departament', 'Library'])
             & (employee['BASE_SALARY']>60000) & (employee['GENDER']=='Female')].head()

POSITION_TITLE  ...    JOB_DATE
249                              STAFF ANALYST  ...  2015-07-18
412                 ADMINISTRATIVE COORDINATOR  ...  2005-04-02
476                  ADMINISTRATIVE SUPERVISOR  ...  2014-11-08
892                              LIBRARIAN III  ...  2002-06-15
1165  DEPUTY ASSISTANT DIRECTOR (EXECUTIVE LEV  ...  2014-03-15

[5 rows x 8 columns]

*Find all the males with salary over 100,000. Return only the race, gender and salary columns*


In [91]:
employee.loc[((employee['GENDER']=='Male') & (employee['BASE_SALARY']>100000)), ['RACE', 'GENDER', 'BASE_SALARY']].head()

RACE GENDER  BASE_SALARY
8                        White   Male     107962.0
11   Black or African American   Male     180416.0
43             Hispanic/Latino   Male     165216.0
169                      White   Male     120916.0
178                      White   Male     210588.0

*Select all salaries as a Series in a separate variable. From this series select all salaries under 25,000*

In [93]:
salaries = employee['BASE_SALARY']
salaries[salaries<25000]

454    24960.0
Name: BASE_SALARY, dtype: float64

*Get the same exact result as exercise 11, but make your selection from the employee DataFrame. Use only a single line of code*

In [95]:
employee.loc[(employee['BASE_SALARY']<25000), ['BASE_SALARY']].head()

BASE_SALARY
454      24960.0